File: 2025-1/sesion-de-clase-08 - Parte 02/SoftProg/SoftProgRRHHDA/src/main/java/pe/edu/pucp/inf30/softprog/rrhh/mysql/BaseDAOImpl.java
package pe.edu.pucp.inf30.softprog.rrhh.mysql;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import pe.edu.pucp.inf30.softprog.config.DBManager;
import pe.edu.pucp.inf30.softprog.rrhh.dao.ICrud;

public abstract class BaseDAOImpl<T> implements ICrud<T> {
    protected abstract PreparedStatement comandoInsertar(Connection conn, T modelo) throws SQLException;
    protected abstract PreparedStatement comandoModificar(Connection conn, T modelo) throws SQLException;
    protected abstract PreparedStatement comandoEliminar(Connection conn, int id) throws SQLException;
    protected abstract PreparedStatement comandoBuscar(Connection conn, int id) throws SQLException;
    protected abstract PreparedStatement comandoListar(Connection conn) throws SQLException;
    
    protected abstract T mapearModelo(ResultSet rs) throws SQLException;
    
    @Override
    public int insertar(T modelo) {
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement cmd = this.comandoInsertar(conn, modelo);
        ) {
            if (cmd.executeUpdate() == 0) {
                System.err.println("El registro no se inserto.");
                return 0;
            }
            
            try (ResultSet rs = cmd.getGeneratedKeys()) {
                return rs.next() ? rs.getInt(1) : -1;
            }
        }
        catch (SQLException e) {
            System.err.println("Error SQL durante la insercion: " + e.getMessage());
            throw new RuntimeException("No se pudo insertar el registro.", e);
        }
        catch (Exception e) {
            System.err.println("Error inpesperado: " + e.getMessage());
            throw new RuntimeException("Error inesperado al insertar el registro.", e);
        }
    }

    @Override
    public boolean modificar(T modelo) {
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement ps = this.comandoModificar(conn, modelo);
        ) {
            return ps.executeUpdate() > 0;
        }
        catch (SQLException e) {
            System.err.println("Error SQL durante la modificacion: " + e.getMessage());
            throw new RuntimeException("No se pudo modificar el registro.", e);
        }
        catch (Exception e) {
            System.err.println("Error inpesperado: " + e.getMessage());
            throw new RuntimeException("Error inesperado al modificar el registro.", e);
        }
    }

    @Override
    public boolean eliminar(int id) {
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement ps = this.comandoEliminar(conn, id);
        ) {
            return ps.executeUpdate() > 0;
        }
        catch (SQLException e) {
            System.err.println("Error SQL durante la eliminacion: " + e.getMessage());
            throw new RuntimeException("No se pudo eliminar el registro.", e);
        }
        catch (Exception e) {
            System.err.println("Error inpesperado: " + e.getMessage());
            throw new RuntimeException("Error inesperado al eliminar el registro.", e);
        }
    }

    @Override
    public T buscar(int id) {
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement ps = this.comandoBuscar(conn, id);
        ) {
            ResultSet rs = ps.executeQuery();
            
            if (!rs.next()) {
                System.err.println("No se encontro el registro con id: " + id);
                return null;
            }
            
            return this.mapearModelo(rs);
        }
        catch (SQLException e) {
            System.err.println("Error SQL durante la busqueda: " + e.getMessage());
            throw new RuntimeException("No se pudo buscar el registro.", e);
        }
        catch (Exception e) {
            System.err.println("Error inpesperado: " + e.getMessage());
            throw new RuntimeException("Error inesperado al buscar el registro.", e);
        }
    }

    @Override
    public List<T> listar() {
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement ps = this.comandoListar(conn);
        ) {
            ResultSet rs = ps.executeQuery();
            
            List<T> modelos = new ArrayList<>();
            while (rs.next()) {
                modelos.add(this.mapearModelo(rs));
            }
            
            return modelos;
        }
        catch (SQLException e) {
            System.err.println("Error SQL durante el listado: " + e.getMessage());
            throw new RuntimeException("No se pudo listar el registro.", e);
        }
        catch (Exception e) {
            System.err.println("Error inpesperado: " + e.getMessage());
            throw new RuntimeException("Error inesperado al listar los registros.", e);
        }
    }
}

File: 2025-1/sesion-de-clase-17/SoftProgNet/SoftProgPersistencia/DAOImpl/BaseDAOImpl.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using PUCP.Edu.Pe.SoftProg.Config;
using PUCP.Edu.Pe.SoftProg.Persistencia.DAO;

namespace PUCP.Edu.Pe.SoftProg.Persistencia.DAOImpl {
    public abstract class BaseDAOImpl<T> : ICrud<T> {
        protected abstract DbCommand CommandoInsertar(DbConnection conn, T modelo);
        protected abstract DbCommand CommandoInsertar(DbConnection conn, DbTransaction transaction, T modelo);
        protected abstract DbCommand CommandoModificar(DbConnection conn, T modelo);
        protected abstract DbCommand CommandoEliminar(DbConnection conn, int id);
        protected abstract DbCommand CommandoBuscar(DbConnection conn, int id);
        protected abstract DbCommand CommandoListar(DbConnection conn);
        protected abstract T mapearModelo(DbDataReader reader);

        protected void AgregarParametroEntrada(DbCommand cmd, string nombre, DbType tipo, object dato) {
            DbParameter param = cmd.CreateParameter();
            param.ParameterName = nombre;
            param.DbType = tipo;
            param.Value = dato;
            param.Direction = ParameterDirection.Input;
            cmd.Parameters.Add(param);
        }

        protected void AgregarParametroSalida(DbCommand cmd, string nombre, DbType tipo) {
            DbParameter param = cmd.CreateParameter();
            param.ParameterName = nombre;
            param.DbType = tipo;
            param.Direction = ParameterDirection.Output;
            cmd.Parameters.Add(param);
        }

        public int Insertar(T modelo) {
            try {
                using (DbConnection conn = DBManager.GetInstance().GetConnection())
                using (DbCommand cmd = this.CommandoInsertar(conn, modelo)) {
                    if (cmd.ExecuteNonQuery() == 0) {
                        Console.Error.WriteLine("El registro no se inserto");
                        return -1;
                    }
                    return (int)cmd.Parameters["@p_id"].Value;
                }
            }
            catch (Exception e) {
                Console.Error.WriteLine("Error SQL Durante la insercion: " + e.Message);
                throw new InvalidOperationException("No se pudo insertar el registro.", e);
            }
        }

        public int Insertar(T modelo, DbConnection conn, DbTransaction transaction) {
            try {
                using (DbCommand cmd = this.CommandoInsertar(conn, transaction, modelo)) {
                    if (cmd.ExecuteNonQuery() == 0) {
                        Console.Error.WriteLine("El registro no se inserto");
                        return -1;
                    }
                    return (int)cmd.Parameters["@p_id"].Value;
                }
            }
            catch (Exception e) {
                Console.Error.WriteLine("Error SQL Durante la insercion: " + e.Message);
                throw new InvalidOperationException("No se pudo insertar el registro.", e);
            }
        }

        public bool Modificar(T modelo) {
            try {
                using (DbConnection conn = DBManager.GetInstance().GetConnection())
                using (DbCommand cmd = this.CommandoModificar(conn, modelo)) {
                    return cmd.ExecuteNonQuery() > 0;
                }
            }
            catch (Exception e) {
                Console.Error.WriteLine("Error SQL Durante la modificacion: " + e.Message);
                throw new InvalidOperationException("No se pudo modificar el registro.", e);
            }
        }

        public bool Eliminar(int id) {
            try {
                using (DbConnection conn = DBManager.GetInstance().GetConnection())
                using (DbCommand cmd = this.CommandoEliminar(conn, id)) {
                    return cmd.ExecuteNonQuery() > 0;
                }
            }
            catch (Exception e) {
                Console.Error.WriteLine("Error SQL Durante la eliminacion: " + e.Message);
                throw new InvalidOperationException("No se pudo eliminar el registro.", e);
            }
        }

        public T Buscar(int id) {
            try {
                using (DbConnection conn = DBManager.GetInstance().GetConnection())
                using (DbCommand cmd = this.CommandoBuscar(conn, id)) 
                using (DbDataReader reader = cmd.ExecuteReader()) { 
                    
                    if (!reader.HasRows) {
                        Console.Error.WriteLine("No se encontro el registro con id: " + id);
                        return default;
                    }

                    reader.Read();
                    return this.mapearModelo(reader);
                }
            }
            catch (Exception e) {
                Console.Error.WriteLine("Error SQL Durante la busueda: " + e.Message);
                throw new InvalidOperationException("No se pudo buscar el registro.", e);
            }
        }

        public List<T> Listar() {
            try {
                using (DbConnection conn = DBManager.GetInstance().GetConnection())
                using (DbCommand cmd = this.CommandoListar(conn))
                using (DbDataReader reader = cmd.ExecuteReader()) {
                    List<T> modelos = new List<T>();

                    while (reader.Read()) { 
                        modelos.Add(this.mapearModelo(reader));
                    }

                    return modelos;
                }
            }
            catch (Exception e) {
                Console.Error.WriteLine("Error SQL Durante la busueda: " + e.Message);
                throw new InvalidOperationException("No se pudo buscar el registro.", e);
            }
        }
    }
}
